### PR TITLE
(PDB-5259) Improve performance of fact-contents queries

### PR DESCRIPTION
**Pull request by rbrw**

### CODE REVIEW

**Comment by austb** [comment]
Initial performance testing shows a significant speed-up for this endpoint. I was using a test database populated with 10,000 nodes of dummy data generated by benchmark.

1. old - https://explain.depesz.com/s/LFQx
1. new - https://explain.depesz.com/s/IArR

To summarize, total query time decreased from 22.1 seconds to 3.5 seconds 🎉 Additionally, the total buffers accessed decreased from `shared hit=13,545,211` to `shared hit=34,553`, a three orders of magnitude decrease.

File: src/puppetlabs/puppetdb/scf/storage_utils.clj [hunk 1]
```diff
@@ -211,6 +225,101 @@
...
+                   (throw (IllegalArgumentException.
+                           (tru "Regex array element wasn't string or integer in {0}"
+                                (pr-str path-rx)))))
+        matches-target (fn [i rx]
+                         (format "\n       and %s[%d] ~ %s"
+                                 column
+                                 (inc i)
+                                 (-> rx ast->db jdbc/single-quote)))]
+    (hcore/raw
+     ;; FIXME: double-check that this should be "="
```
Comment (austb) [inline src/puppetlabs/puppetdb/scf/storage_utils.clj:314]: Should remove this before merging

**Review by austb — APPROVED** [review]

File: src/puppetlabs/puppetdb/cheshire.clj [hunk 2]
```diff
@@ -33,11 +35,17 @@
 (defn add-common-json-encoders!*
   "Non-memoize version of add-common-json-encoders!"
   []
+  ;; The .getPrettyPrinter method on the Jackson jsonGenerator will
+  ;; return nil if `:pretty` is not set as a cheshire option
+  (add-encoder
+   PgArray
+   (fn [^PgArray data ^JsonGenerator generator]
+     ;; If .getArray returns a copy, this approach could be expensive
+     ;; for large values.
```
[the remark this answers is not in the document]
Comment (rbrw) [reply src/puppetlabs/puppetdb/cheshire.clj:42]: I don't think so -- just left that as a note to future selves...

File: src/puppetlabs/puppetdb/scf/storage_utils.clj [hunk 3]
```diff
@@ -211,6 +225,101 @@
...
+               (throw (IllegalArgumentException.
+                       (tru "Path array element wasn't string or integer in {0}"
+                            (pr-str path))))))
+       (apply str)))
+
+(defn ast-rx-path-type-pattern
+  [path]
+  (->> path
+       (map #(cond
+               (string? %) (if (re-matches #"\d+" %) \s \_)
```
[the remark this answers is not in the document]
Comment (rbrw) [reply src/puppetlabs/puppetdb/scf/storage_utils.clj:254]: done

**Review by austb — APPROVED** [review]

**Pull request merged**